container-chains/templates/frontier/runtime/src/precompiles.rs

```diff
@@ -14,71 +14,89 @@
-        [
-            hash(1),
-            hash(2),
-            hash(3),
-            hash(4),
-            hash(5),
-            hash(1024),
-            hash(1025),
-        ]
-    }
-}
-impl<R> PrecompileSet for FrontierPrecompiles<R>
-where
-    R: pallet_evm::Config,
-{
-    fn execute(&self, handle: &mut impl PrecompileHandle) -> Option<PrecompileResult> {
-        match handle.code_address() {
-            // Ethereum precompiles :
-            a if a == hash(1) => Some(ECRecover::execute(handle)),
-            a if a == hash(2) => Some(Sha256::execute(handle)),
-            a if a == hash(3) => Some(Ripemd160::execute(handle)),
-            a if a == hash(4) => Some(Identity::execute(handle)),
-            a if a == hash(5) => Some(Modexp::execute(handle)),
-            // Non-Frontier specific nor Ethereum precompiles :
-            a if a == hash(1024) => Some(Sha3FIPS256::execute(handle)),
-            a if a == hash(1025) => Some(ECRecoverPublicKey::execute(handle)),
-            _ => None,
-        }
+
+    /// Returns the decimals places of the token.
+    fn decimals() -> u8 {
+        18
     }
 
-    fn is_precompile(&self, address: H160, _remaining_gas: u64) -> IsPrecompileResult {
-        IsPrecompileResult::Answer {
-            is_precompile: Self::used_addresses().contains(&address),
-            extra_cost: 0,
-        }
+    /// Must return `true` only if it represents the main native currency of
+    /// the network. It must be the currency used in `pallet_evm`.
+    fn is_native_currency() -> bool {
+        true
     }
 }
 
-fn hash(a: u64) -> H160 {
-    H160::from_low_u64_be(a)
-}
+type EthereumPrecompilesChecks = (AcceptDelegateCall, CallableByContract, CallableByPrecompile);
+
+#[precompile_utils::precompile_name_from_address]
+type TemplatePrecompilesAt<R> = (
+    // Ethereum precompiles:
+    // Allow DELEGATECALL to stay compliant with Ethereum behavior.
+    PrecompileAt<AddressU64<1>, ECRecover, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<2>, Sha256, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<3>, Ripemd160, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<4>, Identity, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<5>, Modexp, EthereumPrecompilesChecks>,
+    // Non-template specific nor Ethereum precompiles :
+    PrecompileAt<AddressU64<1024>, Sha3FIPS256, (CallableByContract, CallableByPrecompile)>,
+    PrecompileAt<AddressU64<1025>, ECRecoverPublicKey, (CallableByContract, CallableByPrecompile)>,
+    // Template specific precompiles:
+    PrecompileAt<
+        AddressU64<2048>,
+        Erc20BalancesPrecompile<R, NativeErc20Metadata>,
+        (CallableByContract, CallableByPrecompile),
+    >,
+    PrecompileAt<
+        AddressU64<2049>,
+        BatchPrecompile<R>,
+        (
+            SubcallWithMaxNesting<2>,
+            // Batch is the only precompile allowed to call Batch.
+            CallableByPrecompile<OnlyFrom<AddressU64<2050>>>,
```

Comment:
```suggestion
            CallableByPrecompile<OnlyFrom<AddressU64<2049>>>,
```

Reply:
Doesn't this have a test?

Reply:
good catch, probably not. @fgamundi can you add a test for this?

Reply:
`CallableByPrecompile` rule removed as it isn't needed for batch to call itself. Batch runs the inner calls without changing the original `caller`, so it doesn't count as coming from a precompile. It's the same case for CallPermit.
Added tests for Batch calling itself and CallPermit calling Batch